File: python/pia/resources/intelligence.py
"""Official PIA SDK - AI Intelligence & Market Analysis Resource."""

from __future__ import annotations

import urllib.parse
from typing import Any, Dict, List, Optional

from ..errors import ValidationError
from ..transport import AsyncTransport, SyncTransport


class IntelligenceResource:
    """Synchronous AI Market Intelligence & Quantitative Analysis resource."""

    def __init__(self, transport: SyncTransport) -> None:
        self._transport = transport

    def analyze(
        self,
        symbol: Optional[str] = None,
        query: Optional[str] = None,
        context: Optional[Dict[str, Any]] = None,
    ) -> Dict[str, Any]:
        """Generates AI-powered real-time quantitative analysis and catalyst explanations."""
        payload: Dict[str, Any] = {}
        if symbol:
            payload["symbol"] = symbol.strip().upper()
        if query:
            payload["query"] = query
        if context:
            payload["context"] = context
        return self._transport.request("/api/v1/intelligence/analyze", method="POST", json_data=payload)

    def get_insights(self, symbol: str) -> Dict[str, Any]:
        """Retrieves synthesized AI narrative insights and price drivers for a specific symbol."""
        if not symbol or not symbol.strip():
            raise ValidationError("Symbol must be a non-empty string.", param_name="symbol")
        clean = symbol.strip().upper()
        encoded = urllib.parse.quote(clean, safe="")
        return self._transport.request(f"/api/v1/market/insights/{encoded}", method="GET")


class AsyncIntelligenceResource:
    """Asynchronous AI Market Intelligence & Quantitative Analysis resource."""

    def __init__(self, transport: AsyncTransport) -> None:
        self._transport = transport

    async def analyze(
        self,
        symbol: Optional[str] = None,
        query: Optional[str] = None,
        context: Optional[Dict[str, Any]] = None,
    ) -> Dict[str, Any]:
        """Asynchronously generates AI-powered quantitative analysis."""
        payload: Dict[str, Any] = {}
        if symbol:
            payload["symbol"] = symbol.strip().upper()
        if query:
            payload["query"] = query
        if context:
            payload["context"] = context
        return await self._transport.request("/api/v1/intelligence/analyze", method="POST", json_data=payload)

    async def get_insights(self, symbol: str) -> Dict[str, Any]:
        """Asynchronously retrieves synthesized AI narrative insights for a specific symbol."""
        if not symbol or not symbol.strip():
            raise ValidationError("Symbol must be a non-empty string.", param_name="symbol")
        clean = symbol.strip().upper()
        encoded = urllib.parse.quote(clean, safe="")
        return await self._transport.request(f"/api/v1/market/insights/{encoded}", method="GET")

File: typescript/src/resources/intelligence.ts
/**
 * Official PIA SDK - AI Intelligence & Market Analysis Resource
 */

import { ValidationError } from "../errors";
import type { HttpTransport } from "../http/transport";
import type { RequestOptions } from "../types";

export interface IntelligenceAnalyzeRequest {
  symbol?: string;
  query?: string;
  text?: string;
  context?: Record<string, any>;
}

export interface IntelligenceAnalyzeResponse {
  symbol?: string;
  sentiment?: "bullish" | "bearish" | "neutral" | string;
  confidence?: number;
  analysis: string;
  catalysts?: string[];
  key_levels?: {
    support?: number[];
    resistance?: number[];
  };
  generated_at: string;
}

export interface MarketInsightResponse {
  symbol: string;
  summary: string;
  sentiment: string;
  drivers?: string[];
  timestamp?: string;
}

export class IntelligenceResource {
  constructor(private readonly transport: HttpTransport) {}

  /**
   * Generates AI-powered real-time quantitative analysis and catalyst explanations.
   * Accepts either a symbol string (e.g. "ETHUSDT") or a structured request payload.
   */
  public async analyze(
    request: string | IntelligenceAnalyzeRequest,
    options?: RequestOptions
  ): Promise<IntelligenceAnalyzeResponse> {
    const payload: IntelligenceAnalyzeRequest =
      typeof request === "string"
        ? { symbol: request, query: request, text: request }
        : request;

    return this.transport.request<IntelligenceAnalyzeResponse>(
      "/api/v1/intelligence/analyze",
      "POST",
      payload,
      options
    );
  }

  /**
   * Retrieves synthesized AI narrative insights and price drivers for a specific symbol.
   */
  public async getInsights(
    symbol: string,
    options?: RequestOptions
  ): Promise<MarketInsightResponse> {
    if (!symbol || typeof symbol !== "string" || symbol.trim() === "") {
      throw new ValidationError("Symbol must be a non-empty string.", "symbol");
    }
    const cleanSymbol = symbol.trim().toUpperCase();
    return this.transport.request<MarketInsightResponse>(
      `/api/v1/market/insights/${encodeURIComponent(cleanSymbol)}`,
      "GET",
      undefined,
      options
    );
  }
}
